### DISCUSSION TITLE
CustomOrthographicController

### DISCUSSION THREAD

**Discussion by argmichael** [visgl/deck.gl#6413] [Q&A] [answered]
Hey guys, 

I have a question about Custom Orthographic Controller. 
So I have my main DeckGL component here:
```
 <DeckGL
        id="deck-gl-main-component"
        style={DECKGL_CSS_STYLE}
        layers={layers}
        viewState={newViewState as ViewStateProps}
        controller={controller() as any}
        ...
>
```
I have my controller function which is wrapped in a usecallback here:
```
  const controller = useCallback(
    () =>
      getCustomOrthographicController(
        keybinds,
        ...
      ),
    [keybinds]
  );
```
Inside getCustomOrthographicController:
```
export function getCustomOrthographicController(
  keybinds: CustomKeyBinds,
) {
  class CustomOrthographicController extends OrthographicController {
    handleEvent(event: ControllerEvent) {
      /* "Switch" */

      const {
        selectNote,
        addNotes1,
        addNotes2,
        addNotes3,
        ruler,
      } = keybinds;

      ...
    }
  }

  return CustomOrthographicController;
}

```
Keybinds is an object that contains key and value pair where key is the action and the value is a letter:
```
{selectNote: "A"
addNotes1: "I"
addNotes2: "P"
addNotes3: "Q"
ruler: "R"
}
```
My question is, in my main component where it contains my controller function, when keybinds change the useCallback knows that keybinds have changed and the values have been updated but inside my getCustomOrthographicController, the values never get updated, it always take the original values when my web_app starts. For example, if keybinds has selectNote originally with a value of T and I change it to a value of O. In my main component it changes but my getCustomOrthographicController still holds the original value of keybinds which has selectNote with a value of T. Any help would be appreciated. Thank you!

**Comment by Pessimistress** [comment, marked as answer]
Controller is only created once for each view. To update keybindings in your use case, you should pass options like this:

```jsx
const [keybindings, setKeybindings] = useState({});

<DeckGL ...
controller={{
  type: CustomOrthographicController,
  keybindings
}}
/>
```

```js
class CustomOrthographicController extends OrthographicController {
    setProps(props) {
        this.keybindings = props.keybindings;
        super.setProps(props);
    }

    handleEvent(event: ControllerEvent) {
      const {
        selectNote,
        addNotes1,
        addNotes2,
        addNotes3,
        ruler,
      } = this.keybinds;

      ...
    }
  }
```
